tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+    /// too little data in the buffer of the named pipe.
+    ///
+    /// You can play around with the [`named-pipe-peek` example] to get a feel
+    /// for how this function behaves.
+    ///
+    /// [total_bytes_available]: PipePeekInfo::total_bytes_available
+    /// [`named-pipe-peek` example]: https://github.com/tokio-rs/tokio/blob/master/examples/named-pipe-peek.rs
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::io;
+    /// use tokio::io::{AsyncReadExt as _, AsyncWriteExt as _};
+    /// use tokio::net::windows::{NamedPipeOptions, NamedPipeClientOptions};
+    ///
+    /// const PIPE_NAME: &str = r"\\.\pipe\tokio-named-pipe-peek-consumed";
+    /// const N: usize = 100;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let mut server = NamedPipeOptions::new().create(PIPE_NAME)?;
+    /// let mut client = NamedPipeClientOptions::new().create(PIPE_NAME)?;
+    /// server.connect().await?;
+    ///
+    /// let client = tokio::spawn(async move {
+    ///     for _ in 0..N {
+    ///         client.write_all(b"ping").await?;
+    ///     }
+    ///
+    ///     let mut buf = [0u8; 4];
+    ///     client.read_exact(&mut buf).await?;
+    ///
+    ///     Ok::<_, io::Error>(buf)
+    /// });
+    ///
+    /// let mut buf = [0u8; 4];
+    /// let mut available = 0;
+    ///
+    /// for n in 0..N {
+    ///     if available < 4 {
+    ///         server.read_exact(&mut buf).await?;
+    ///         assert_eq!(&buf[..], b"ping");
+    ///
+    ///         let info = server.peek(None)?;
+    ///         available = info.total_bytes_available;
+    ///         continue;
+    ///     }
+    ///
+    ///     // here we know that at least `available` bytes are immediately
+    ///     // ready to read.
+    ///     server.read_exact(&mut buf).await?;
+    ///     available -= buf.len();
+    ///     assert_eq!(&buf[..], b"ping");
+    /// }
+    ///
+    /// server.write_all(b"pong").await?;
+    ///
+    /// let buf = client.await??;
+    /// assert_eq!(&buf[..], b"pong");
+    /// # Ok(()) }
+    /// ```
+    pub fn peek(&mut self, mut buf: Option<&mut ReadBuf<'_>>) -> io::Result<PipePeekInfo> {
+        use std::convert::TryFrom as _;
+
+        unsafe {
+            let mut n = mem::MaybeUninit::zeroed();
+            let mut total_bytes_available = mem::MaybeUninit::zeroed();
+            let mut bytes_left_this_message = mem::MaybeUninit::zeroed();
+
+            let result = {
+                let (buf, len) = match &mut buf {
+                    Some(buf) => {
+                        let len = DWORD::try_from(buf.capacity())
+                            .expect("buffer too large for win32 api");
+                        // Safety: the OS has no expectation on whether the
+                        // buffer is initialized or not.
+                        let buf = buf.inner_mut() as *mut _ as *mut _;
+                        (buf, len)
```

Comment:
This will overwrite the part of the `ReadBuf` that is already marked as filled. You should use `unfilled_mut` instead.
```suggestion
                        let unfilled = buf.unfilled_mut();
                        let len = DWORD::try_from(unfilled.len())
                            .expect("buffer too large for win32 api");
                        // Safety: the OS has no expectation on whether the
                        // buffer is initialized or not.
                        (unfilled.as_mut_ptr() as LPVOID, len)
```

Reply:
Too long since I've used `ReadBuf`. Cheers!